include/boost/math/fft/real_algorithms.hpp

```diff
@@ -0,0 +1,149 @@
+///////////////////////////////////////////////////////////////////
+//  Copyright Eduardo Quintana 2021
+//  Copyright Janek Kozicki 2021
+//  Copyright Christopher Kormanyos 2021
+//  Distributed under the Boost Software License,
+//  Version 1.0. (See accompanying file LICENSE_1_0.txt
+//  or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_FFT_REAL_ALGORITHMS_HPP
+  #define BOOST_MATH_FFT_REAL_ALGORITHMS_HPP
+
+  #include <algorithm>
+  #include <numeric>
+  #include <cmath>
+  
+  namespace boost { namespace math {  namespace fft {
+  
+  namespace detail {
+  
+  template<class RealType>
+  RealType complex_root_of_unity_real(long n,long p=1)
+  /*
+    Computes cos(-2 pi p/n)
+  */
+  {
+    p = modulo(p,n);
+    
+    if(p==0)
+      return RealType(1);
+    
+    long g = gcd(p,n); 
+    n/=g;
+    p/=g;
+    switch(n)
+    {
+      case 1:
+        return RealType(1);
+      case 2:
+        return p==0 ? RealType(1) : RealType(-1);
+      case 4:
+        return p==0 ? RealType(1) : 
+               p==1 ? RealType(0) :
+               p==2 ? RealType(-1) :
+                      RealType(0) ;
+    }
+    using std::cos;
+    RealType phase = -2*p*boost::math::constants::pi<RealType>()/n;
+    return RealType(cos(phase));
+  }
+  template<class RealType>
+  RealType complex_root_of_unity_imag(long n,long p=1)
+  /*
+    Computes sin(-2 pi p/n)
+  */
+  {
+    p = modulo(p,n);
+    
+    if(p==0)
+      return RealType(0);
+    
+    long g = gcd(p,n); 
+    n/=g;
+    p/=g;
+    switch(n)
+    {
+      case 1:
+        return RealType(0);
+      case 2:
+        return p==0 ? RealType(0) : RealType(0);
+      case 4:
+        return p==0 ? RealType(0) : 
+               p==1 ? RealType(-1) :
+               p==2 ? RealType(0) :
+                      RealType(1) ;
+    }
+    using std::sin;
+    RealType phase = -2*p*boost::math::constants::pi<RealType>()/n;
+    return RealType(sin(phase));
+  }
+  
+  template <class T>
+  void real_dft_power2(const T *in_first, const T *in_last, T* out, int sign)
```

Comment:
Here is the real-to-halfcomplex rfft for powers of 2. Changing the sign of the transform has the only effect of reversing the sign of the imaginary part in the result.

Reply:
> real-to-halfcomplex rfft for powers of 2

Cool. I'll give it a try as soon as i get a chance. Thanks Eduardo!